BUG: 1.12.0rc1: build failure on windows due to unintentional macro usage in boost

It looks like boost is using an unfortunate choice of template variable name on windows, which conflicts with a macro:
```
In file included from ..\scipy\_lib\boost_math\include\boost/math/policies/policy.hpp:11:
..\scipy\_lib\boost_math\include\boost/math/tools/mp.hpp(171,34): error: expected parameter declarator
  171 | template<typename L, std::size_t I>
      |                                  ^
C:\Program Files (x86)\Windows Kits\10\include\10.0.22621.0\ucrt\complex.h(63,20): note: expanded from macro 'I'
   63 | #define I          _Complex_I
      |                    ^
C:\Program Files (x86)\Windows Kits\10\include\10.0.22621.0\ucrt\complex.h(62,29): note: expanded from macro '_Complex_I'
   62 | #define _Complex_I _FCbuild(0.0F, 1.0F)
      |                             ^
```

From https://github.com/conda-forge/scipy-feedstock/pull/262, using the vendored boost.

CC @rgommers @tylerjereddy

So with `cpp_std=c++17` and the updated https://github.com/scipy/scipy/pull/19761 that now completely disables `aligned_alloc` on windows, things look much better (compiles, no crashes), only leaving two very minor tolerance violations:
```
=========================== short test summary info ===========================
FAILED stats/tests/test_distributions.py::TestGenExpon::test_sf_isf[0.125-0.9508674287164001-0.25-5-0.5] - AssertionError: 
Not equal to tolerance rtol=1e-14, atol=0

Mismatched elements: 1 / 1 (100%)
Max absolute difference: 1.55431223e-15
Max relative difference: 1.24344979e-14
 x: array(0.125)
 y: array(0.125)
FAILED stats/tests/test_distributions.py::TestGenExpon::test_cdf_ppf[0.125-0.04913257128359998-0.25-5-0.5] - AssertionError: 
Not equal to tolerance rtol=1e-14, atol=0

Mismatched elements: 1 / 1 (100%)
Max absolute difference: 1.55431223e-15
Max relative difference: 1.24344979e-14
 x: array(0.125)
 y: array(0.125)
= 2 failed, 59028 passed, 2542 skipped, 259 xfailed, 14 xpassed, 7 warnings in 1033.56s (0:17:13) =
```

> only leaving two very minor tolerance violations:

Spoke too soon, without AVX512 we get two more; a variant of https://github.com/scipy/scipy/issues/15533, as well as:
```
FAILED optimize/tests/test__differential_evolution.py::TestDifferentialEvolutionSolver::test_L1 - AssertionError: 
Not equal to tolerance rtol=1e-07, atol=0.0005

Mismatched elements: 1 / 13 (7.69%)
Max absolute difference: 0.0005003
Max relative difference: 0.00019409
 x: array([0.999956, 0.999957, 0.99996 , 0.999969, 0.999907, 0.999966,
       0.999856, 0.999926, 0.999909, 2.999611, 2.9995  , 2.999572,
       0.999806])
 y: array([1, 1, 1, 1, 1, 1, 1, 1, 1, 3, 3, 3, 1])
```
unfortunately, 2e-4 for the relative tolerance is pretty bad (if not unacceptable) IMO.

--- a/meson.build
+++ b/meson.build
@@ -11,7 +11,7 @@ project(
     'buildtype=debugoptimized',
     'b_ndebug=if-release',
     'c_std=c99',
-    'cpp_std=c++14',
+    'cpp_std=c++17',
     'fortran_std=legacy',
     'blas=openblas',
     'lapack=openblas'

--- a/scipy/io/_fast_matrix_market/meson.build
+++ b/scipy/io/_fast_matrix_market/meson.build
@@ -27,7 +27,6 @@ py3.extension_module('_fmm_core',
     'src/_fmm_core_write_coo_32.cpp',
     'src/_fmm_core_write_coo_64.cpp',
   ],
-  override_options: ['cpp_std=c++17'],
   cpp_args: [
     '-DFMM_SCIPY_PRUNE',
     '-DFMM_USE_FAST_FLOAT',

